### DISCUSSION TITLE
Export Workflow built with builder to json for import to another instance.

### DISCUSSION THREAD

**Discussion by tunechr** [elsa-workflows/elsa-core#1487] [Q&A]
Hi

I know it seems a little odd but I wanted to find the best way of building our workflows and then being able to import to other instances.  

Out workflow has around 150 steps so doing this in the UI is a bit long-winded. 
The code-based Builder is a lot quicker, but I don't know the best way to pull this out so I can then load it into another instance (tenant).

Any ideas?

**Comment by sfmskywalker** [comment]
I probably misunderstood the question completely, but if you create a workflow with the Workflow Builder API, isn't your workflow automatically part of all of your instances?

The same goes for workflows created in the designer: they are available to all of your instances. If by "instance" you mean a node in a cluster.

But like I said, I probably misunderstood so please let me know.

**Reply by tunechr** [reply]
Hi 

Sorry, It's my explanation which isn't very good!
I'm not 100% across the definitions/blueprints/instances and how they are related.

I have a number of workflows, one of which is around 120 activities in total. 
Building this with the UI is a bit slow and using the Builder in code is a lot quicker for me to update and test, so being able to do it in code and export it would be great to keep it as a version and be able to import to a users tenant.  

I can see that the workflows defined in code appear in the register, but I can't export them to create a new version?

Another option would be for me to have the code built workflow as a standard workflow and allow users to use it as a base or modify this in some way.  So I was hoping there could be a way of exporting this or using it as a base for a new definition.

Hope that makes sense?

Chris

**Reply by sfmskywalker** [reply]
I understand now. You want to create workflows in code, export them from the Registry so that you can save them as JSON files and for example import them into other workflow server installations.

Unfortunately this is not supported. The main reason this is hard to do is because when you define a workflow in code, you are free to use C# lambda expressions, which do not automatically translate to actual C# statements that then get compiled again.

Coded workflows and workflows created from the designer are two sources that get translated into a workflow blueprint (which is what Elsa uses at runtime to execute workflows). This is unidirectional, meaning that you cannot go back from a workflow blueprint to a workflow definition or coded workflow.

I'm not saying that's impossible to do, just that it's hard to do (and not implemented in Elsa).

With Elsa 3 it would become easier due to the change in object model of a workflow graph, but that's neither here nor there.

What you can try to do is build a workflow blueprint serializer, bearing in mind that C# expressions won't be (easily) convertible to e.g. JS expressions. But you could monkey-patch this later on manually if you want.

Alternatively, you could consider writing your workflow directly in JSON format, which is directly importable into the designer. Although I doubt that this would be a faster process than using the designer. 
You could of course start with the designer, export the JSON and then continue hand-crafting the JSON file, but this can be tricky with all the UUID elements.

What I think you really need is a clean workflow DSL that's neither C# nor JSON which allows you to code workflows quickly and easily and that can translate into JSON (which would also allow you to directly import it in the designer if you wanted to).

There's an [issue for that](https://github.com/elsa-workflows/elsa-core/issues/462), but it's not yet planned for implementation.

**Reply by tunechr** [reply]
Yes, exactly that!
Thanks so much for the answer, that give me some great insights into the inner workings. 
I'm going to go for writing in the JSON format, I think that's the best solution for now. 

Just wanted to say thanks to you and the team working on this project, It's really great to see the developments, It's really helped our project so far and allowed us to keep things in the .net space. 

Chris
